#319 / create a notification context

vdedios opened the pull request:
closes #319 

Here is the idea:
1) we have a shared notification context to display push message for users
2) we wrap all the routes with the context so notifications are available across the app
3) we only have to use `notify(msg)` and `warn(msg)` to display messages to the user

vdedios commented on webapp/src/index.tsx:
```diff
@@ -44,97 +45,99 @@ root.render(
-            }
-          />
-          {developmentMode && (
+        <NotificationContextProvider>
```
As said, we wrap the whole app with the context

vdedios commented on webapp/src/pages/EditAvatarPage/EditAvatarPage.tsx:
```diff
@@ -70,35 +63,28 @@ export default function EditAvatarPage() {
-          type: 'success',
-          message: 'Image visible area saved correctly.',
-        });
+        notify('Image visible area saved correctly.');
```
And now we only need to call a function hook, no need to manage notification states on each component

gbudau replied:
praise: Thanks for updating these pages!

vdedios commented on webapp/src/shared/context/NotificationContext.tsx:
```diff
@@ -0,0 +1,82 @@
+  const [notification, setNotification] = React.useState<Notification>();
+  const [visible, setVisible] = React.useState<boolean>(false);
+
+  const notify = (message: string) => {
+    if (message.length) {
+      setNotification({ message, type: 'info', position: 'top' });
+      setVisible(true);
+    }
+  };
+
+  const warn = (message: string) => {
+    if (message.length) {
+      setNotification({ message, type: 'warning', position: 'bottom' });
+      setVisible(true);
+    }
```
Since this was new and didn't have a design for this I kind of made this up:
- informative notifications -> will display in the top of the page with green color (like the text we had)
- warning notifications -> will display in the bottom of the page with yellow color (like the text we had)

I'm not particularly in love with this decision, we could have both appear from bottom and use top notifications for new future types of notifications we want to display. Or maybe use another colors, I don't mind.

gbudau commented on webapp/src/shared/components/Snackbar/Snackbar.css:
```diff
@@ -8,14 +10,14 @@
   min-height: 3rem;
-  min-width: 15rem;
+  min-width: 18rem;
   color: var(--color-light);
```
thought: The white text on yellow/green background is a bit hard to read.
![yellow-white](https://user-images.githubusercontent.com/54190400/201090999-7acf4b35-cc13-4c88-81ab-a53a4617b969.png)
![green-white](https://user-images.githubusercontent.com/54190400/201091016-031588d3-f398-4c47-8b86-2f06c0cabf6c.png)

question: What do you think about changing this to `color: var(--color-background);`?
![yellow-black](https://user-images.githubusercontent.com/54190400/201091059-1e5ad8df-4df4-4089-800e-2209352b9fb6.png)

![green-black](https://user-images.githubusercontent.com/54190400/201091069-c3471ca5-37d7-40aa-b6a4-8590fddca421.png)


Source: https://accessibleweb.com/color-contrast-checker/

vdedios replied:
Yes! nice idea :+1:

gbudau commented on webapp/src/pages/ComponentsBookPage/ComponentsExamples/SnackBarExample.tsx:
```diff
@@ -1,40 +1,28 @@
         <Button
           variant={ButtonVariant.SUBMIT}
-          onClick={() => setVisibleTop((prev) => !prev)}
+          onClick={() => notify('This is an informative notification!')}
```
praise: I love the new API :clap:

gbudau commented on webapp/src/pages/RegisterPage/RegisterPage.tsx:
```diff
@@ -91,31 +74,28 @@ export default function RegisterPage() {
+
+      notify('Registration complete');
+
       setTimeout(() => {
         navigate(LOGIN_OPTIONS_URL);
       }, 2000);
```
```suggestion
      navigate(LOGIN_OPTIONS_URL);
```

suggestion (if-minor): We could remove the timeout because this notification persists even if the page changes :heart:

vdedios replied:
Nice idea!! done

gbudau commented on webapp/src/shared/components/Input/EditUserForm.tsx:
```diff
@@ -40,31 +33,37 @@ export default function EditUserForm({
   const [status, setStatus] = useState<EditUserFormSubmitStatus>(
     initialSubmitFormStatus,
   );
```
```suggestion
```

suggestion (if-minor): We could remove this `useState`, the `useEffect`, and the types associated with it because we don't use them anymore.

gbudau reviewed (approved):
Thanks! I love how you implemented the context and API :heart:

vdedios commented:
@gbudau requesting a review of yours again because I've updated this branch with main and added this notifications to the chatroom page, in case anything is wrong. Thanks again!

(I've also added the `warn` and `notify` functions inside a `useCallback` since they could be used inside effects, so we won't to avoid extra renders)

gbudau commented on webapp/src/pages/ChatroomPage/ChatroomPage.tsx:
```diff
@@ -37,10 +31,7 @@ type ChatroomProps = {
-    type: 'pending',
-    message: '',
-  });
+  const { warn } = useNotificationContext();
```
praise: Thank you! Looks great!

suggestion: There are some CSS rules we don't use anymore:

<details>

```diff
diff --git a/webapp/src/pages/ChatroomPage/ChatroomPage.css b/webapp/src/pages/ChatroomPage/ChatroomPage.css
index 1184e70..d8bac03 100644
--- a/webapp/src/pages/ChatroomPage/ChatroomPage.css
+++ b/webapp/src/pages/ChatroomPage/ChatroomPage.css
@@ -4,20 +4,15 @@
   background-color: var(--color-background);
   color: var(--color-light);
   display: grid;
+  gap: 1rem;
   grid-template-areas:
     'header'
     'messages'
-    'message-input'
-    'status';
-  grid-template-rows: auto 1fr auto auto;
+    'message-input';
+  grid-template-rows: auto 1fr auto;
   grid-template-columns: 1fr;
 }
 
-.chatroom-header,
-.chatroom-messages {
-  margin-bottom: 1rem;
-}
-
 .chatroom-loading {
   grid-column: 1/-1;
   grid-row: 1/-1;
diff --git a/webapp/src/pages/LoginPage/LoginPage.css b/webapp/src/pages/LoginPage/LoginPage.css
index e5637fe..bc278a0 100644
--- a/webapp/src/pages/LoginPage/LoginPage.css
+++ b/webapp/src/pages/LoginPage/LoginPage.css
@@ -60,11 +60,6 @@
   width: 100%;
 }
 
-.login-form > p:empty::before {
-  content: '';
-  display: inline-block;
-}
-
 .final-text {
   display: flex;
   justify-content: center;
diff --git a/webapp/src/pages/RegisterPage/RegisterPage.css b/webapp/src/pages/RegisterPage/RegisterPage.css
index 8eacdf1..091d06f 100644
--- a/webapp/src/pages/RegisterPage/RegisterPage.css
+++ b/webapp/src/pages/RegisterPage/RegisterPage.css
@@ -47,11 +47,6 @@
   max-width: 20rem;
 }
 
-.register-form .inputs-container > p:empty::before {
-  content: '';
-  display: inline-block;
-}
-
 .register .landing-animation {
   display: none;
 }

```

</details>

vdedios replied:
done! thanks for the suggestions

gbudau reviewed (approved):
Thank you! LGTM :+1:

fizcamposancos reviewed (approved):
Very nice. I really like it
